### ISSUE TITLE
Bug: [no-unnecessary-condition] false positive on indexed access of a generic type

### ISSUE THREAD

**Issue by mx-bernhard** [typescript-eslint/typescript-eslint#7850]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&fileType=.ts&code=GYVwdgxgLglg9mABAUwB4EMC2AHANsgHgBUVUpkwATAZ0TgCMArZaAPgAoAoRRdAJwDmALkREANN0QBrZAE8RM2XGCienAJQiAbnBiVEAb0kQE1KInoBGRAF5eggNqKAuogCEdsCFy4A3MdNzegAmW3sBJzlXAH5oxAByYHQfenQIKXj-AF9OEzAzOhEDRGA4OBEzPhgwAQAfLx9ELLDi0vLEBtwm-zQsPGR2ODEEtvj1XyA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg10NM8AOXapUAYQAW6aAGsylDt160isAKpF2JdABMACgENMh%2BOjxYFYAGaHUOnjShrN2vQBlY0O1dv3lTpAuRNDScnoAkkS66AAeegCC0NDoJFZKvAC%2BIJlAA&tokens=false

### Repro Code

```typescript
function example<T extends object>(
  arg: T,
  key: keyof T  
): void {
  const b1 = arg[key] != null;
  const b2 = arg[key] ?? 'fallback';
}
const o: { foo: string|null } = { foo: null };
example(o, 'foo');
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

No error should be reported since it is possible to supply null values on properties of the incoming object in the example.

### Actual Result

An error gets reported saying that null and the value of the type that it is compared with does not have any overlap:

```
Unnecessary conditional, the types have no overlap. 5:14 - 5:30

Unnecessary conditional, expected left-hand side of `??` operator to be possibly null or undefined. 6:14 - 6:22
```
